src/Avalonia.Controls/Presenters/TextPresenter.cs

```diff
@@ -818,11 +839,23 @@ protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
             base.OnAttachedToVisualTree(e);
 
             _caretTimer.Tick += CaretTimerTick;
+
+            if (_canvas == null)
+            {
+                _canvas = new TextSelectionCanvas();
+            }
+
+            _layer = TextSelectorLayer.GetTestLayer(this);
+            _layer?.Add(_canvas);
+            _canvas?.SetPresenter(this);
         }
 
         protected override void OnDetachedFromVisualTree(VisualTreeAttachmentEventArgs e)
         {
             base.OnDetachedFromVisualTree(e);
+            if (_canvas is { } c)
+                _layer?.Remove(c);
+            _canvas?.SetPresenter(null);
```

Comment:
Nit: can be move inside the above if since it's already making the null check, e.g. `c.SetPresenter(null)`